### PR TITLE
Refactor the code for feature transformation and label construction

### PR DESCRIPTION
**Pull request by zheng-da**
*Description of changes:*
This PR refactors the code for feature transformation and label construction. The rules are formally explained in README.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/transform.py [hunk 1]
```diff
@@ -255,12 +282,152 @@ def process_labels(data, label_confs, is_node):
         train_mask[train_idx] = 1
         val_mask[val_idx] = 1
         test_mask[test_idx] = 1
-    if label_col is None:
-        return {'train_mask': train_mask,
-                'val_mask': val_mask,
-                'test_mask': test_mask}
+        if self.label_name is not None:
+            train_mask_name = self.label_name + '_train_mask'
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:286]: We need to change GFS code for customized mask names.

File: python/graphstorm/gconstruct/transform.py [hunk 2]
```diff
@@ -255,12 +282,152 @@ def process_labels(data, label_confs, is_node):
...
+            All data associated with nodes/edges of a node/edge type.
+
+        Returns
+        -------
+        dict of Tensors : it contains the labels as well as training/val/test splits.
+        """
+        assert self.col_name in data, f"The label column {self.col_name} does not exist."
+        label = data[self.col_name]
+        assert np.issubdtype(label.dtype, np.integer) \
+                or np.issubdtype(label.dtype, np.floating), \
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:318]: Can it be a float?
There is a conflict between the check and the error message.

File: python/graphstorm/gconstruct/transform.py [hunk 3]
```diff
@@ -133,31 +181,19 @@ def process_features(data, ops):
...
-        else:
-            new_data[feat_name] = data[feat_col]
+    for op in ops:
+        res = op(data[op.col_name])
+        assert isinstance(res, dict)
+        for key, val in res.items():
+            new_data[key] = val
     return new_data
 
 def get_valid_label_index(label):
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:199]: Can we treat negative integer, i.e., -1 as invalid labels for integer label data?

File: python/graphstorm/gconstruct/transform.py [hunk 4]
```diff
@@ -255,12 +282,152 @@ def process_labels(data, label_confs, is_node):
...
+        data : dict of Tensors
+            All data associated with nodes/edges of a node/edge type.
+
+        Returns
+        -------
+        dict of Tensors : it contains training/val/test splits for link prediction.
+        """
+        # Any column in the data can define the number of samples in the data.
+        assert len(data) > 0, "The edge data is empty."
+        for val in data.values():
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:374]: Why not just num_samples = len(data[0])

File: python/graphstorm/gconstruct/transform.py [hunk 5]
```diff
@@ -255,12 +282,152 @@ def process_labels(data, label_confs, is_node):
...
+    assert 'task_type' in label_conf, "'task_type' must be defined in the label field."
+    task_type = label_conf['task_type']
+    # By default, we use all labels for training.
+    split_pct = label_conf['split_pct'] if 'split_pct' in label_conf else [1, 0, 0]
+
+    if task_type == 'classification':
+        assert 'label_col' in label_conf, \
+                "'label_col' must be defined in the label field."
+        label_col = label_conf['label_col']
+        return [ClassificationProcessor(label_col, label_col, split_pct)]
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:406]: We need to think about the case when a user provides valid and test set.

**Review by classicsong — COMMENTED** [review]
One general question is: If we provide customizable naming of train, val and test masks and customizable naming of text token related features, how to modify GraphStorm to support it?
We need to add more input arguments to allow users to specify the names of certain special fields. We also need to modify the DataLoader and HF support to accommodate the changes.

File: python/graphstorm/gconstruct/README.md [hunk 6]
```diff
@@ -1,34 +1,69 @@
...
+is stored in a dictionary. A node dictionary contains multiple fields and
+most fields are optional:
+* `node_type` specifies the node type. This field is mandatory.
+* `files` specifies the input files for the node data. This field is mandatory.
+There are multiple options to specify the input files.
+For a single input file, it contains the path of a single file.
+For multiple files, it contains the path of files with a wildcard,
+or a list of file paths.
+* `format` specifies the input file format. This field is mandatory.
+Currently, the pipeline supports two formats: parquet and JSON.
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/README.md:22]: we don't support CSV right now.

File: python/graphstorm/gconstruct/README.md [hunk 7]
```diff
@@ -1,34 +1,69 @@
...
+The detailed format information is specified in the format section.
+* `node_id_col` specifies the column that contains the node IDs. This field is optional.
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/README.md:24]: if two dicts provide the node features, one of the dicts should contain the node IDs while the other doesn't have node IDs. otherwise, we need to check if the two node IDs are consistent. There is a check in the code to verify if every node type has an ID mapping generated from the node IDs in the input file. If a node type doesn't have node IDs, it will report an error.

File: python/graphstorm/gconstruct/README.md [hunk 8]
```diff
@@ -1,34 +1,69 @@
...
+are listed in the section of Feature Transformation.
+
+A label dictionary is defined:
+* `task_type` specifies the task defined on the nodes or edges. The field is mandatory.
+Currently, its value can be `classification`, `regression` and `link_prediction`.
+* `label_col` specifies the column name in the input file that contains the label.
+This has to be specified for classification and regression tasks.
+`label_col` is used as the label name.
+* `split_pct` specifies how to split the data into training/validation/test.
+This is optional. If it's not specified, all data will be used for training.
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/README.md:60]: we can change the default behavior. What is the practice in Neptune ML?

File: python/graphstorm/gconstruct/README.md [hunk 9]
```diff
@@ -1,34 +1,69 @@
...
+The pipeline constructs three additional vectors indicating
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/README.md:61]: we should. we need to add additional fields to support that.

[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/README.md:61]: there is this issue here https://github.com/awslabs/graphstorm/issues/41, but i haven't implemented this yet. we need to discuss what interface we should provide.

File: python/graphstorm/gconstruct/README.md [hunk 10]
```diff
@@ -98,3 +131,24 @@ python3 -m graphstorm.gconstruct.construct_graph \
...
+Currently, the graph construction pipeline supports two input formats: Parquet and JSON.
+
+For the Parquet format, each column defines a node/edge feature, label or node/edge IDs.
+For multi-dimensional features, currently the pipeline requires the features to be stored
+as a list of vectors. The pipeline will reconstruct multi-dimensional features and store
+them in a matrix.
+
+For JSON format, each line of the JSON file is a JSON object. The JSON object can only
+have one level. The value of each field can only be primitive values, such as integers,
+strings and floating points, or a list of primitive values.
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/README.md:145]: we don't. i'll update it.

File: python/graphstorm/gconstruct/construct_graph.py [hunk 11]
```diff
@@ -311,13 +312,16 @@ def process_edge_data(process_confs, node_id_map, convert2ext_mem,
         in_files = get_in_files(process_conf['files'])
         feat_ops = parse_feat_ops(process_conf['features']) \
                 if 'features' in process_conf else None
+        label_ops = parse_label_ops(process_conf['labels'], is_node=False) \
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:315]: yes, they need to create a label object, but they don't need to specify the label column in the label field.

File: python/graphstorm/gconstruct/transform.py [hunk 12]
```diff
@@ -51,35 +84,48 @@ def __call__(self, strs):
         att_masks = []
         type_ids = []
         for s in strs:
+            assert isinstance(s, str), "The input of the tokenizer has to be a string."
             t = self.tokenizer(s, max_length=self.max_seq_length,
                                truncation=True, padding='max_length', return_tensors='pt')
             tokens.append(t['input_ids'])
             # The masks are small integers. We can use int4 or int8 to store them.
             # This can signficantly reduce memory consumption.
             att_masks.append(t['attention_mask'].to(th.int8))
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:93]: true. let's introduce another PR to fix it.

File: python/graphstorm/gconstruct/transform.py [hunk 13]
```diff
@@ -51,35 +84,48 @@ def __call__(self, strs):
...
             type_ids.append(t['token_type_ids'].to(th.int8))
-        return {'token_ids': th.cat(tokens, dim=0).numpy(),
-                'attention_mask': th.cat(att_masks, dim=0).numpy(),
-                'token_type_ids': th.cat(type_ids, dim=0).numpy()}
+        if self.feat_name is not None:
+            token_id_name = self.feat_name + '_token_ids'
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:96]: if we don't do so, we need to report errors if users provide two text features.

File: python/graphstorm/gconstruct/transform.py [hunk 14]
```diff
@@ -99,28 +144,31 @@ def parse_feat_ops(confs):
...
         if 'transform' not in feat:
-            transform = None
+            transform = Noop(feat['feature_col'], feat_name)
         else:
-            transform = feat['transform']
-            assert 'name' in transform, \
-                    "'name' must be defined in the transformation field."
-            if transform['name'] == 'tokenize_hf':
-                transform = parse_tokenize(transform)
+            conf = feat['transform']
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:159]: we need to update parse_feat_ops to parse the new transform field.

File: python/graphstorm/gconstruct/transform.py (hunk 3 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:199]: currently, we only treat NaN as invalid label values.

File: python/graphstorm/gconstruct/transform.py (hunk 5 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:406]: let's address this in another PR.

File: python/graphstorm/gconstruct/transform.py (hunk 4 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:374]: data is a dict. we cannot use `[0]` to index it.

**Comment by zheng-da** [comment]
I revert the change of using different names to store different token tensors and different labels.
Currently, we can only allow one text feature and one label for each node/edge type.

**Review by classicsong — APPROVED** [review]
LGTM
